draft-irtf-cfrg-hpke.md

```diff
@@ -428,37 +486,36 @@ def KeySchedule(mode, pkR, zz, enc, info, psk, pskID, pkSm):
   VerifyMode(mode, psk, pskID, pkSm)
 
   pkRm = Marshal(pkR)
-  identifier = "RFCXXXX"
   ciphersuite = concat(encode_big_endian(kem_id, 2),
                        encode_big_endian(kdf_id, 2),
                        encode_big_endian(aead_id, 2))
-  pskID_hash = Hash(pskID)
-  info_hash = Hash(info)
-  context = concat(identifier, ciphersuite, mode, enc, pkRm,
+  pskID_hash = LabeledExtract(0, "pskID", pskID)
+  info_hash = LabeledExtract(0, "info", info)
+  context = concat(ciphersuite, mode, enc, pkRm,
                    pkSm, pskID_hash, info_hash)
 
-  secret = Extract(psk, zz)
-  key = Expand(secret, concat("key", context), Nk)
-  nonce = Expand(secret, concat("nonce", context), Nn)
-  exporter_secret = Expand(secret, concat("exp", context), Nh)
+  if len(psk) > Nb then:
```

Comment:
Why do we need this explicit hashing step?  Shouldn't the internals of `LabeledExtract` handle that for us?

Reply:
Are you suggesting this if should be inside LabeledExtract? This is the only place where we give an arbitrary-length bitstring as the first parameter to (Labeled)Extract, so maybe it's wasteful to have the branching every time (even for all the calls with 0 as first parameter)? If you say you prefer this to have a cleaner KeySchedule, I think I could agree.

Just in case: Have you seen my earlier comment? It might have been buried within this long discussion, so let me cite:

https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50#issuecomment-603568796

> On the PSK length:
> 
> HMAC(K, M) feeds K through a call to Hash if K is longer than Hash's blocksize. If we allow arbitrary-size PSKs for user convenience, this would introduce another place where Hash is called on arbitrary-sized input, just like with info and pskID. (As a reminder: Arbitrary-sized input to Hash can lead to collisions with input to Hash calls inside HMAC. This would render the proof inapplicable that proves indifferentiability of HMAC from a random oracle. To avoid collisions, we use Extract calls and separate the input domains by using a prefix).
> 
> To avoid this extra Hash call inside HMAC, we can just anticipate it directly in KeySchedule, and thus control the prefix. While this introduces branching inside KeySchedule, this branching would either way be done inside HMAC. Doing the Extract call even if the PSK is shorter then Nb would hurt performance in non-PSK modes, too.

Reply:
I am reading specs of Keccak and Blake2/3 to see if they do internal hashing in their KDF/XOF modes that could collide with a Hash call inside HPKE. This `len > Nb` step is very specific to HMAC, and people might implement Extract and Expand with something different. So maybe we can come up with a more general statement here instead of hardcoding support for HMAC.

Reply:
Right, I was thinking that in the HKDF case, this duplicated what HMAC was doing internally.  If there's nothing current that is doing something that would collide, I would suggest we remove this, and maybe add a note about the assumption we're making about the KDF (no internal hashing that could collide).

Reply:
@bifurcation this isn't quite duplicating what HMAC is doing internally, since it explicitly prefixes the input before hashing.

Reply:
With commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/421828db601168b9f81acc6d299754e477e4d10a, the PSK is now always fed through LabeledExtract. I included my reasoning in the commit message. Please let me know if that makes sense to you.

Reply:
An alternative to feeding the PSK always through LabeledExtract is to restrict the PSK to a fixed length, for example to Nh.

Reply:
👍

Reply:
If someone implements Extract and Expand with something else than HKDF, like a modern hash function's KDF and XOF modes, then this PSK hashing step is probably not needed. I attempted a formulation with commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/09acc770c7604185836085499b7284dcfa0568a8.